Manual Trigger user Audit

loghen41 opened the pull request:
This is in response in part to https://github.com/concourse/concourse/issues/2752

It only solves half of the issue, specifically focusing on manual triggers.

On a manual trigger, the user_name of the client is appended to the web log.
This serves as a start for an audit trail for manual triggers

Signed-off-by: Logan Hendricks <llhendricks@xactware.com>

loghen41 commented:
@pivotal-jwinters Thanks for the review.  To be honest, I put together this pull request, and then after submitting it, I was looking at some of the upcoming pull requests your team has put together and I truthfully did wonder if my changes would fit into your future roadmap.

I actually had the same thought along the lines that you did. I did wonder if the accessor really had enough of the needed information to log out a proper audit log.  (and it seemed somewhat more secure for the accessor itself to log the info, rather than the external modules to have to request the needed audit data from the accessor and log it out themselves).

My only concern is that while for auditing you really want everything logged out that is being accomplished, it can be considered TOO verbose if it's just a general log for all requests that the accessor receives. Do you have any insights into that before I jump into it?

pivotal-jwinters commented:
I think we can probably log everything to start and see how bad it is. If we have a flag in the atc to turn auditing on/off then people can use it as needed. 

If its too chatty then maybe we let people configure which `actions` they want to audit?

loghen41 commented:
@pivotal-jwinters Sounds good. I got a bit stalled at the end of last week, but I'll be picking it up again today and see what I can accomplished

loghen41 commented:
@pivotal-jwinters I just finished a very general audit logger.  Sorry that took me a bit longer than expected.  I've attached it to the DEBUG level of logging. (cloudfoundry.lager didn't have an option for anything between DEBUG and INFO... so debug seemed more appropriate for that level of verbosity.)

Please feel free to review, and let me know if there are anything you feel I may have missed.

pivotal-jwinters commented:
@loghen41 it looks the unit tests are failing.

you'll have to `go generate ./...` in `atc/api` to regenerate fakes.

loghen41 commented:
k, I'll go take care of that.  I thought I caught all of them, but must have missed one.

pivotal-jwinters commented on atc/api/accessor/handler.go:
```diff
@@ -3,31 +3,38 @@ package accessor
 	ctx := context.WithValue(r.Context(), "accessor", acc)
 
+	if acc, ok := acc.(Access); ok {
+		h.logger.Debug("audit", lager.Data{"action": r.URL.Path, "method": r.Method, "user": acc.UserName(), "parameters": r.URL.Query()})
```
I think logging the full path and method might be too generic. [If you look at the way we set up routes](https://github.com/concourse/concourse/blob/master/atc/routes.go#L128), each one has a name (`GetJob`) and path params (e.g. `:team_name`). 

In the accessorHandler the name gets injected as `h.action` and the path params are available as `r.FormValues` (example [here](https://github.com/concourse/concourse/blob/master/atc/api/jobserver/get.go#L14)). They might also be available as query params but i'm not sure about that.

What if we logged just the `h.action` and path params along with the `UserName`. Most of our api requests don't supply many traditional query params, but we could potentially include those as well.

Thoughts?

loghen41 replied:
@pivotal-jwinters That sounds great to me. I'm grateful to get some feedback on it. as I was looking for something a bit more precise than just stating the full path and method.  

I had dived through the different attributes available, but must have missed those.  (especially as this is my first time looking through a lot of your codebase)

pivotal-jwinters commented:
@jama22 you may be interested in this

@vito any opinions on how much we would be logging here

cirocosta commented:
cc @scottietremendous 

*update: as an improvement, I just created https://github.com/concourse/concourse/issues/3404 to cover the lack of `go:generate` instructions in our contributing guide*

loghen41 commented:
Sorry to disappear off of the face of the earth for a week or so. The last week was somewhat crazy on my end.  

@pivotal-jwinters I implemented the changes you recommended by just logging the action and the FormValues. You were correct in that the way you folks transform your requests, this is a more accurate representation of what is being requested.

After finishing those initial changes, I decided to do what you recommended earlier, and created cli arguments that can be passed to the web agent (or docker container) that will Enable/Disable logging events for a particular group of actions.  I broke them down as best as I could see fit into 9 different categories:

```Builds
Containers
Jobs
Pipelines
Resources
System
Team
Workers
Volumes```

All of the routable actions are mapped to a particular category, and turned off or on based on the flags passed at Web start time.

Lastly, in that conversion to allowing CLI arguments, I moved all of the logic related to audting to it's own package, and built a series of unit tests surrounding the new changes.

I will say, all of this is incredibly open to review and change.  So, if you folks don't like any of it, Please feel free to make it known.  In particular, I would love to find a better way to improve [this](https://github.com/loghen41/concourse/blob/580894963bd080f555eda38576dd4fefa17afe1e/atc/api/accessor/handler.go#L36l).  I was working with a previous [unit test](https://github.com/loghen41/concourse/blob/580894963bd080f555eda38576dd4fefa17afe1e/atc/api/accessor/handler_test.go#L59), which had an expectation of being able to provide a nil accessor object... which works fine in the source, but not in the test suite.

But just know that none of this is sacred... and also hopefully some of it is helpful.

pivotal-jwinters commented:
I'm really liking the changes, just a couple of things:

1. Yeah [this](https://github.com/loghen41/concourse/blob/580894963bd080f555eda38576dd4fefa17afe1e/atc/api/accessor/handler.go#L36l) seems unnecessary. In your tests you can inject a fake so you don't have to deal with nils.

1. Small nit, can we change `audit` -> `auditor`? Realistically we should have called `access` -> `accessor` as well, but thats a different discussion lol.

1. Can you rebase your changes. Your PR has a bunch of commits from our master branch.

1. A bunch of fakes got regenerated. It looks like you might be using a different version of counterfeiter than us (no idea who's is newer lol). But can you make sure to only stage the new fakes, so we don't have so many files changed.

loghen41 commented:
@pivotal-jwinters Thanks for the quick response.

I'll go review that particular unit test.

That's fine with me. I'll change that to `auditor`

as for `Can you rebase your changes. Your PR has a bunch of commits from our master branch.` I can definitely do that. I actually realized I was merging for most of it, and my last commit was a rebase instead.  My apologies to clog up the git history.

And sorry for the update of the fakes.  I should have just reverted some of them.  I had implemented one version of this commit, and then backed out of it when I broke it up into a separate package.  I'll be sure to watch that in the future.

I'll notify once I get a new commit in.

loghen41 commented:
@pivotal-jwinters I believe this pull request is now in a finished state.  There is a failure in the testflight, but looking at your build history, I was wondering if that is something that is failing off and on by itself potentially due to the 5.0 release? You'll have to let me know.

After following a rebase as you recommended... it appears to be that even more of your teams' commits were brought into my forked repo.  Which was kinda the opposite of what we both wanted.  

I'm open to whatever would be easiest for you folks.  I can attempt to clean up the git history on my end, we can ignore the extra commits, or I could go to the far extreme of blowing away this PR, Reforking the Repo, and cherry picking over the needed changes.

Also, if there are further changes you want me to make, let me know.

Lastly, I've been using this comment system to keep in touch, would it be better to just connect with you on the discord channel you have for the future?

pivotal-jwinters commented:
@loghen41 yeah we're basically at a point where I can't even tell what your changes are anymore lol

If you could squash all your changes into a single commit, and have that sitting atop master that would be great. You could cherry pick them over to a new branch, as you say, or try an interactive rebase on the existing branch.

loghen41 commented:
@pivotal-jwinters I agree. this PR is basically garbage for git history.  Let me see if I can squash them, or do an interactive rebase. I'll avoid cherry picking things over as a last ditch effort.  

Sorry again for the level of mess that it's become, I've personally never had this happen before, so I was surprised when it kept getting worse.

pivotal-jwinters commented:
haha no worries

loghen41 commented:
@pivotal-jwinters   After looking at squashing my commits, and doing an interactive rebase... I found that the amount of unneeded and extra commits from my account were pervasive back to almost 180 commits... which seemed far to invasive to try to squash.

One of my main problems I found on this particular Pull Request is that I was working on my master branch of my fork, which means that it's pretty much impossible to reset my files appropriately here.  I'll be blowing away my original fork, cherry picking over the changes, and re-creating a pull request (which sadly seems the easiest of all of the options)

I'll close this pull request soon, and open the new one this afternoon.  Sorry again for all of the issues.
